### PR TITLE
[rllib] MB MPO

### PR DESCRIPTION
**Pull request by michaelzhiluo**
MB-MPO Agent

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/agents/mbmpo/mbmpo.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+    # Metaupdate Step with outer combine loop for multiple MAML iterations
+    train_op = rollouts.combine(
+        MetaUpdate(workers, config["num_maml_steps"],
+                   config["maml_optimizer_steps"], metric_collect))
+    return train_op
+
+
+def get_policy_class(config):
+    # @mluo: TODO
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:302]: Can we get a ValueError here in case framework != "torch".

Comment (sven1977) [reply rllib/agents/mbmpo/mbmpo.py:302]: Probably best in `validate_config` at the top.

File: rllib/agents/mbmpo/mbmpo.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    "inner_lr": 1e-3,
+    # Horizon of Environment (200 in MB-MPO paper)
+    "horizon": 200,
+    # Dynamics Ensemble Hyperparameters
+    "dynamics_model": {
+        "custom_model": DynamicsEnsembleCustomModel,
+        # Number of Transition-Dynamics Models for Ensemble
+        "model_ensemble_size": 1,
+        # Hidden Layers for Model Ensemble
+        "model_hiddens": [512, 512],
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:71]: Can we call this `fcnet_hiddens`, same as in most other model configs.

File: rllib/agents/mbmpo/mbmpo.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    "maml_optimizer_steps": 8,
+    # Inner Adaptation Step size
+    "inner_lr": 1e-3,
+    # Horizon of Environment (200 in MB-MPO paper)
+    "horizon": 200,
+    # Dynamics Ensemble Hyperparameters
+    "dynamics_model": {
+        "custom_model": DynamicsEnsembleCustomModel,
+        # Number of Transition-Dynamics Models for Ensemble
+        "model_ensemble_size": 1,
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:69]: I think we should cut the "model_" prefix. It's kind of redundant.

File: rllib/agents/mbmpo/mbmpo.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+# yapf: enable
+
+# Select Metric Keys for MAML Stats Tracing
+METRICS_KEYS = [
+    "episode_reward_mean", "episode_reward_min", "episode_reward_max"
+]
+
+
+class MetaUpdate:
+    def __init__(self, workers, num_steps, maml_steps, metric_gen):
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:100]: Please add a detailed docstring as to what this exec-related method does and which args it takes in the c'tor.

File: rllib/agents/mbmpo/mbmpo.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        self.workers = workers
+        self.num_steps = num_steps
+        self.step_counter = 0
+        self.maml_optimizer_steps = maml_steps
+        self.metric_gen = metric_gen
+        self.metrics = {}
+
+    def __call__(self, data_tuple):
+        # Metaupdate Step
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:109]: docstring

File: rllib/agents/mbmpo/mbmpo.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+            self.workers.sync_weights()
+            return []
+
+    def postprocess_metrics(self, metrics, prefix=""):
+        for key in metrics.keys():
+            self.metrics[prefix + "_" + key] = metrics[key]
+
+
+def post_process_metrics(prefix, workers, metrics):
+    # Obtain Current Dataset Metrics and filter out
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:169]: docstring

File: rllib/agents/mbmpo/mbmpo.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+    for i, e in enumerate(workers.remote_workers()):
+        e.learn_on_batch.remote(samples[i])
+
+
+def fit_dynamics(policy, pid):
+    return policy.dynamics_model.fit()
+
+
+def sync_ensemble(workers, model_attr="dynamics_model"):
+    def get_ensemble_weights(worker):
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:187]: docstring

File: rllib/agents/mbmpo/mbmpo.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        set_func = ray.put(set_normalizations)
+        for e in workers.remote_workers():
+            e.foreach_policy.remote(
+                set_func, normalizations=normalization_dict)
+
+
+# Similar to MAML Execution Plan
+def execution_plan(workers, config):
+    # Train TD Models
+    workers.local_worker().foreach_policy(fit_dynamics)[0]
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo.py:236]: why do we need the [0] index here? It's not assigned to anything.

File: rllib/agents/mbmpo/mbmpo_torch_policy.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    policy.dynamics_model = ModelCatalog.get_model_v2(
+        obs_space,
+        action_space,
+        num_outputs=num_outputs,
+        model_config=config["dynamics_model"],
+        framework="torch",
+        name="dynamics_ensemble",
+    )
+
+    action_dist, num_outputs = ModelCatalog.get_action_dist(
```
Comment (sven1977) [inline rllib/agents/mbmpo/mbmpo_torch_policy.py:36]: Actually, you can skip this now and just use `build_torch_policy(make_model=...)` (w/o the `_and_action_dist`, iff the action dist is a default one anyways.

File: rllib/tuned_examples/mbmpo/halfcheetah-mbmpo.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+halfcheetah-mb-mpo:
+    env: ray.rllib.examples.env.halfcheetah.HalfCheetahWrapper
+    run: MBMPO
+    stop:
+        training_iteration: 500
+    local_dir: ~/mbmpo-testing
```
Comment (sven1977) [inline rllib/tuned_examples/mbmpo/halfcheetah-mbmpo.yaml:6]: do we need this dir here?

File: rllib/tuned_examples/mbmpo/halfcheetah-mbmpo.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    config:
+        framework: torch
```
Comment (sven1977) [inline rllib/tuned_examples/mbmpo/halfcheetah-mbmpo.yaml:8]: Add comment: # only supported in torch right now

File: rllib/examples/env/halfcheetah.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    if "args" in locals:
+        args += locals["args"]
+
+    if "kwargs" in locals:
+        kwargs_dict.update(locals["kwargs"])
+    return args, kwargs_dict
+
+
+class HalfCheetahWrapper(HalfCheetahEnv):
+    def __init__(self, *args, **kwargs):
```
Comment (sven1977) [inline rllib/examples/env/halfcheetah.py:22]: Please document, what this does and why we need a wrapper here.

File: rllib/agents/mbmpo/model_vector_env.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
+import logging
```
Comment (sven1977) [inline rllib/agents/mbmpo/model_vector_env.py:1]: We should move this to `rllib/env/.`. It's not necessarily specific to MBMPO.

File: rllib/agents/mbmpo/model_vector_env.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+from ray.rllib.utils.annotations import override
+from ray.rllib.env.vector_env import VectorEnv
+from ray.rllib.evaluation.rollout_worker import get_global_worker
+
+logger = logging.getLogger(__name__)
+
+
+def custom_model_vector_env(**kwargs):
+    worker_index = kwargs["env_context"].worker_index
```
Comment (sven1977) [inline rllib/agents/mbmpo/model_vector_env.py:12]: docstring

File: rllib/agents/mbmpo/model_vector_env.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+    @override(VectorEnv)
+    def vector_step(self, actions):
+        if self.cur_obs is None:
+            raise ValueError("Need to reset env first")
+
+        obs_batch = np.stack(self.cur_obs, axis=0)
+        action_batch = np.stack(actions, axis=0)
+
+        next_obs_batch = self.model.predict_model_batches(
```
Comment (sven1977) [inline rllib/agents/mbmpo/model_vector_env.py:71]: I'm not sure I understand. Why do we only predict the next-obs-batch using the very first TD-Model (index [0] in c'tor gets stored in self.model).

File: rllib/agents/mbmpo/model_vector_env.py (hunk 13 above)
Comment (sven1977) [reply rllib/agents/mbmpo/model_vector_env.py:1]: Wondering, why we need this at all. Can't we just use the already existing vector wrapping functionality? Should the actually wrapped ModelEnv not expose the same API as any Env wrapped as a VectorEnv? In other words, the logic in this class here to call `self.model` to get the next obs (inside `vector_step`) should be abstracted inside the single ModelEnv, not here in the VectorizerWrapper. Let me know, if I'm not seeing something.

**Review by sven1977 — CHANGES_REQUESTED** [review]
This is great! Just smaller things, mostly missing docstrings. These are really important especially for the distributed exec API classes, so that users can follow what the algo is doing.

File: rllib/agents/mbmpo/model_vector_env.py (hunk 15 above)
Comment (michaelzhiluo) [reply rllib/agents/mbmpo/model_vector_env.py:71]: Each worker using only 1 TD-model. This index is selected ahead of time (random.int). It is in the forward method in  model_ensemble.py. The index [0] is just that worker.foreach_policy returns a list!

File: rllib/tuned_examples/mbmpo/halfcheetah-mbmpo.yaml (hunk 10 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/mbmpo/halfcheetah-mbmpo.yaml:6]: Will be deleted

File: rllib/agents/mbmpo/model_vector_env.py (hunk 13 above)
Comment (michaelzhiluo) [reply rllib/agents/mbmpo/model_vector_env.py:1]: Your first comment: I agree! This will be moved in the next commit, as soon as I add improved ensemble training and MAML dataset batching (like PPO minibatching), which hopefully result in identical performance to the original codebase. This will likely be committed tmrw.

Your second comment: It is more efficient to wrap it around a set of environments, rather than one enviornment. This is because it is not as efficient to call your policy `num_env` times versus 1 simple call to a batch of (obs, actions) from vec_env. This is a lot faster that way and is also done in the original codebase.

File: rllib/agents/mbmpo/mbmpo.py (hunk 8 above)
Comment (michaelzhiluo) [reply rllib/agents/mbmpo/mbmpo.py:266]: Foreach_policy returns a list but yeah dont need it

File: rllib/agents/mbmpo/model_vector_env.py (hunk 13 above)
Comment (michaelzhiluo) [reply rllib/agents/mbmpo/model_vector_env.py:1]: Actually I will keep this in MBMPO rn, since it is hard to identify another MB algorithm that uses the specifics indicated in this file. (Dreamer is different)

**Comment by michaelzhiluo** [comment]
@ericl Do you have time today to review this PR? Most files are linter and can be ignored.

**Comment by ericl** [comment]
@michaelzhiluo please fix the LINT errors: https://travis-ci.com/github/ray-project/ray/jobs/365290862 , and ensure tests are passing.

File: rllib/agents/ddpg/ddpg_tf_policy.py [hunk 16]
```diff
@@ -296,10 +295,12 @@ def gradients_fn(policy, optimizer, loss):
         clip_func = tf.identity
 
     # Save grads and vars for later use in `build_apply_op`.
-    policy._actor_grads_and_vars = [
-        (clip_func(g), v) for (g, v) in actor_grads_and_vars if g is not None]
-    policy._critic_grads_and_vars = [
-        (clip_func(g), v) for (g, v) in critic_grads_and_vars if g is not None]
+    policy._actor_grads_and_vars = [(clip_func(g), v)
```
Comment (ericl) [inline rllib/agents/ddpg/ddpg_tf_policy.py:298]: You probably have the wrong YAPF version installed. Please revert these changes so the PR is not changing random other files.

**Review by ericl — CHANGES_REQUESTED** [review]
Comments inline

**Review by sven1977 — APPROVED** [review]
Nice work Michael! Thanks for the new algo :)
Only two things:
- LINT needs to pass before we can merge
- test_dependency_torch still fails b/c of some hard dependency (instead of a soft one) in your MBMPO model.

**Comment by michaelzhiluo** [comment]
@ericl @sven1977 Passes Lint Test now, and got rid of the excess files.

File: rllib/agents/mbmpo/model_ensemble.py [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+        if output_nonlinearity:
+            self.layers.append(output_nonlinearity())
+
+        self.model = nn.Sequential(*self.layers)
+
+    def forward(self, x):
+        return self.model(x)
+
+
+class TDDataset(torch.utils.data.Dataset):
```
Comment (sven1977) [inline rllib/agents/mbmpo/model_ensemble.py:61]:
```suggestion
if torch:
    class TDDataset(torch.utils.data.Dataset):
```

File: rllib/agents/mbmpo/model_ensemble.py [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+            delta = normalize(delta, norms["delta"])
+
+        self.x = np.concatenate([obs, actions], axis=1)
+        self.y = delta
+
+    def __len__(self):
+        return self.count
+
+    def __getitem__(self, index):
+        return self.x[index], self.y[index]
```
Comment (sven1977) [inline rllib/agents/mbmpo/model_ensemble.py:80]:
```suggestion
if torch:
    class TDDataset(torch.utils.data.Dataset):
        def __init__(self, dataset: SampleBatchType, norms):
            self.count = dataset.count
            obs = dataset[SampleBatch.CUR_OBS]
            actions = dataset[SampleBatch.ACTIONS]
            delta = dataset[SampleBatch.NEXT_OBS] - obs

            if norms:
                obs = normalize(obs, norms[SampleBatch.CUR_OBS])
                actions = normalize(actions, norms[SampleBatch.ACTIONS])
                delta = normalize(delta, norms["delta"])

            self.x = np.concatenate([obs, actions], axis=1)
            self.y = delta

        def __len__(self):
            return self.count

        def __getitem__(self, index):
            return self.x[index], self.y[index]
```

**Comment by michaelzhiluo** [comment]
@ericl  I think this PR is good to go

**Comment by sven1977** [comment]
@ericl Could you approve? All tests are green now. Thx!

**Comment by sven1977** [comment]
@michaelzhiluo

File: rllib/agents/mbmpo/mbmpo.py [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+        self.metric_gen = metric_gen
+        self.metrics = {}
+
+    def __call__(self, data_tuple):
+        """Arguments:
+            data_tuple (tuple): 1st element is samples collected from MAML
+            Inner adaptation steps and 2nd element is accumulated metrics
+        """
+        # Metaupdate Step
+        print("Meta-Update Step")
```
Comment (ericl) [inline rllib/agents/mbmpo/mbmpo.py:127]: Use logger.info()

**Review by ericl — APPROVED** [review]
LGTM but please replace print() with logging
